File: backend/analyzer/kinematics.py
"""Kinematic metrics: Range-of-Motion and tremor (velocity/acceleration).

The CTR-GCN model gives us a *qualitative* signal — "does this look like
good form?" — but clinicians and physios want *quantitative* numbers
that translate directly to recovery milestones. This module is the
geometric counterpart to the learned model, computed from the same
sliding-window buffer of normalised keypoints:

* :func:`joint_angle_series` — per-frame angle (in degrees) of the
  triplet ``(a, vertex, b)``; e.g. the knee angle from hip → knee → ankle.
* :func:`range_of_motion` — peak-to-trough swing of an angle series.
  This is the canonical clinical ROM number.
* :func:`tremor_metrics` — RMS of velocity and acceleration of a 1D
  signal. Smooth motion → ~0; jitter → large numbers. Cheap, robust to
  scale, and matches what the rehab literature reports as "movement
  smoothness".

All functions accept numpy arrays so they're trivially composable with
the analyzer's existing buffer.
"""
from __future__ import annotations

from typing import Dict

import numpy as np

# Triplets used to define each clinical joint angle. The vertex is the
# joint whose angle we report; the other two are the limb segments
# meeting at it.
JOINT_ANGLE_TRIPLETS: Dict[str, tuple[str, str, str]] = {
    "left_knee":   ("left_hip",      "left_knee",      "left_ankle"),
    "right_knee":  ("right_hip",     "right_knee",     "right_ankle"),
    "left_elbow":  ("left_shoulder", "left_elbow",     "left_wrist"),
    "right_elbow": ("right_shoulder","right_elbow",    "right_wrist"),
    "left_hip":    ("left_shoulder", "left_hip",       "left_knee"),
    "right_hip":   ("right_shoulder","right_hip",      "right_knee"),
}


def joint_angle(a: np.ndarray, vertex: np.ndarray, b: np.ndarray) -> float:
    """Angle at ``vertex`` in degrees, between rays vertex→a and vertex→b.

    Returns ``nan`` if either segment has zero length (a "missing"
    landmark situation), letting callers ignore the frame without
    polluting downstream statistics with a fake 0.
    """
    v1 = np.asarray(a, dtype=np.float64) - np.asarray(vertex, dtype=np.float64)
    v2 = np.asarray(b, dtype=np.float64) - np.asarray(vertex, dtype=np.float64)
    n1 = np.linalg.norm(v1)
    n2 = np.linalg.norm(v2)
    if n1 < 1e-9 or n2 < 1e-9:
        return float("nan")
    cos_theta = np.clip(np.dot(v1, v2) / (n1 * n2), -1.0, 1.0)
    return float(np.degrees(np.arccos(cos_theta)))


def joint_angle_series(
    window_TVC: np.ndarray,
    landmark_names: tuple[str, ...],
    triplet: tuple[str, str, str],
) -> np.ndarray:
    """Compute the angle at ``triplet[1]`` for every frame in the window.

    Frames where any required landmark is missing produce ``nan`` so
    downstream consumers can call ``np.nanmax`` / ``np.nanmin`` safely.
    """
    if window_TVC.ndim != 3:
        raise ValueError(f"window_TVC must be (T, V, C), got {window_TVC.shape}")
    ia = landmark_names.index(triplet[0])
    iv = landmark_names.index(triplet[1])
    ib = landmark_names.index(triplet[2])
    T = window_TVC.shape[0]
    out = np.empty(T, dtype=np.float64)
    for t in range(T):
        out[t] = joint_angle(window_TVC[t, ia], window_TVC[t, iv], window_TVC[t, ib])
    return out


def range_of_motion(angle_series: np.ndarray) -> Dict[str, float]:
    """Min / max / range for an angle series, ignoring NaNs."""
    if angle_series.size == 0 or np.all(np.isnan(angle_series)):
        return {"min_deg": float("nan"), "max_deg": float("nan"), "range_deg": float("nan")}
    mn = float(np.nanmin(angle_series))
    mx = float(np.nanmax(angle_series))
    return {"min_deg": mn, "max_deg": mx, "range_deg": mx - mn}


def tremor_metrics(signal: np.ndarray) -> Dict[str, float]:
    """RMS of velocity and acceleration of a 1D signal.

    For a perfectly smooth sinusoid sampled densely the discrete first
    and second differences are O(1/T) and O(1/T²) respectively — close
    enough to 0 that the tests use a small absolute threshold rather
    than exact equality.
    """
    if signal.size < 3:
        return {"velocity_rms": 0.0, "acceleration_rms": 0.0}
    sig = signal.astype(np.float64)
    # Ignore NaN segments by linear-interpolating across them; this keeps
    # the derivatives well-defined when a frame is briefly missing.
    if np.isnan(sig).any():
        idx = np.arange(sig.size)
        mask = np.isnan(sig)
        if mask.all():
            return {"velocity_rms": 0.0, "acceleration_rms": 0.0}
        sig[mask] = np.interp(idx[mask], idx[~mask], sig[~mask])
    v = np.diff(sig)
    a = np.diff(v)
    return {
        "velocity_rms": float(np.sqrt(np.mean(v * v))),
        "acceleration_rms": float(np.sqrt(np.mean(a * a))),
    }


__all__ = [
    "JOINT_ANGLE_TRIPLETS",
    "joint_angle",
    "joint_angle_series",
    "range_of_motion",
    "tremor_metrics",
]

File: ios/DPMonitor/Core/Kinematics.swift
//
//  Kinematics.swift
//  DPMonitor
//
//  Geometric counterpart to the learned model. Exact port of
//  backend/analyzer/kinematics.py.
//
//  CTR-GCN answers "does this look like good form?"; clinicians also want
//  numbers that map onto recovery milestones. These are computed from the
//  same sliding window:
//
//    * jointAngle        — angle at a vertex, in degrees
//    * rangeOfMotion     — peak-to-trough swing (the clinical ROM number)
//    * tremorMetrics     — RMS of velocity and acceleration; smooth motion
//                          tends to 0, jitter grows
//
//  Everything runs in `Double` because the Python reference is float64 and
//  the parity tests hold to 1e-4. NaN is a first-class value here: a
//  missing landmark must propagate as NaN, never silently become 0, or it
//  would drag a patient's reported ROM toward zero.
//

import Accelerate
import Foundation
import simd

enum Kinematics {

    // MARK: - Joint angle

    /// Angle at `vertex` in degrees, between the rays vertex→a and vertex→b.
    ///
    /// Returns `.nan` when either segment has (near) zero length — the
    /// "missing landmark" case. Callers use `isNaN` to skip the frame.
    static func jointAngle(a: SIMD3<Double>,
                           vertex: SIMD3<Double>,
                           b: SIMD3<Double>) -> Double {
        let v1 = a - vertex
        let v2 = b - vertex
        let n1 = simd_length(v1)
        let n2 = simd_length(v2)
        guard n1 >= 1e-9, n2 >= 1e-9 else { return .nan }
        let cosTheta = min(max(simd_dot(v1, v2) / (n1 * n2), -1.0), 1.0)
        return acos(cosTheta) * 180.0 / Double.pi
    }

    /// Convenience overload for the `Float` storage used by the pipeline.
    static func jointAngle(a: SIMD3<Float>,
                           vertex: SIMD3<Float>,
                           b: SIMD3<Float>) -> Double {
        jointAngle(a: SIMD3<Double>(a), vertex: SIMD3<Double>(vertex), b: SIMD3<Double>(b))
    }

    /// Read the angle for one triplet out of a flat 33×3 frame.
    static func jointAngle(frame: [Float],
                           triplet: (name: String, a: Int, vertex: Int, b: Int)) -> Double {
        @inline(__always)
        func joint(_ v: Int) -> SIMD3<Double> {
            SIMD3(Double(frame[v * 3]), Double(frame[v * 3 + 1]), Double(frame[v * 3 + 2]))
        }
        return jointAngle(a: joint(triplet.a), vertex: joint(triplet.vertex), b: joint(triplet.b))
    }

    /// Every clinical angle for one frame. NaN entries are dropped, matching
    /// `CTRGCNAnalyzer._latest_frame_angles`.
    static func frameAngles(_ frame: [Float]) -> [String: Double] {
        var out = [String: Double](minimumCapacity: PoseLandmarks.angleTriplets.count)
        for triplet in PoseLandmarks.angleTriplets {
            let angle = jointAngle(frame: frame, triplet: triplet)
            if !angle.isNaN { out[triplet.name] = angle }
        }
        return out
    }

    /// Per-frame angle series over a `(T, V, C)` window.
    /// Frames with a missing landmark yield NaN so `rangeOfMotion` can skip them.
    static func jointAngleSeries(window: [[Float]],
                                 triplet: (name: String, a: Int, vertex: Int, b: Int)) -> [Double] {
        window.map { jointAngle(frame: $0, triplet: triplet) }
    }

    // MARK: - Range of motion

    struct ROM {
        var minDeg: Double
        var maxDeg: Double
        var rangeDeg: Double

        static let undefined = ROM(minDeg: .nan, maxDeg: .nan, rangeDeg: .nan)
        var isDefined: Bool { !minDeg.isNaN && !maxDeg.isNaN }
    }

    /// Min / max / range of an angle series, ignoring NaNs.
    static func rangeOfMotion(_ series: [Double]) -> ROM {
        guard !series.isEmpty else { return .undefined }
        var mn = Double.infinity
        var mx = -Double.infinity
        var sawValue = false
        for v in series where !v.isNaN {
            sawValue = true
            if v < mn { mn = v }
            if v > mx { mx = v }
        }
        guard sawValue else { return .undefined }
        return ROM(minDeg: mn, maxDeg: mx, rangeDeg: mx - mn)
    }

    // MARK: - Tremor

    struct Tremor {
        var velocityRMS: Double
        var accelerationRMS: Double

        static let zero = Tremor(velocityRMS: 0, accelerationRMS: 0)
    }

    /// RMS of the first and second differences of a 1-D signal.
    ///
    /// NaN runs are linearly interpolated first so the derivatives stay
    /// defined across a brief dropout. A series shorter than 3 samples, or
    /// one that is entirely NaN, reports zeros rather than NaN — that keeps
    /// the fatigue variance well-defined for very short reps.
    static func tremorMetrics(_ signal: [Double]) -> Tremor {
        guard signal.count >= 3 else { return .zero }

        var sig = signal
        if sig.contains(where: { $0.isNaN }) {
            guard interpolateNaNsInPlace(&sig) else { return .zero }
        }

        var sumV = 0.0
        var sumA = 0.0
        let nV = sig.count - 1
        let nA = sig.count - 2

        var prevV = 0.0
        for i in 0..<nV {
            let v = sig[i + 1] - sig[i]
            sumV += v * v
            if i > 0 {
                let a = v - prevV
                sumA += a * a
            }
            prevV = v
        }

        return Tremor(velocityRMS: (sumV / Double(nV)).squareRoot(),
                      accelerationRMS: nA > 0 ? (sumA / Double(nA)).squareRoot() : 0)
    }

    /// Replicates `np.interp` over NaN positions, including its edge
    /// behaviour: samples before the first valid point take that point's
    /// value, samples after the last take the last point's value.
    ///
    /// - Returns: `false` when every sample is NaN (nothing to interpolate from).
    @discardableResult
    static func interpolateNaNsInPlace(_ sig: inout [Double]) -> Bool {
        let validIndices = sig.indices.filter { !sig[$0].isNaN }
        guard let first = validIndices.first, let last = validIndices.last else {
            return false
        }

        for i in sig.indices where sig[i].isNaN {
            if i <= first {
                sig[i] = sig[first]
            } else if i >= last {
                sig[i] = sig[last]
            } else {
                // Bracket i between the nearest valid samples on each side.
                // `validIndices` is sorted, so a linear scan from the front
                // is fine for the 64-sample windows we deal with.
                var lo = first
                var hi = last
                for v in validIndices {
                    if v < i { lo = v } else { hi = v; break }
                }
                let span = Double(hi - lo)
                let t = span > 0 ? Double(i - lo) / span : 0
                sig[i] = sig[lo] + (sig[hi] - sig[lo]) * t
            }
        }
        return true
    }

    // MARK: - Window summary

    struct JointSummary {
        var joint: String
        var rom: ROM
        var tremor: Tremor
    }

    /// ROM + tremor for every tracked joint over a `(T, V, C)` window.
    ///
    /// Computed on the UN-normalised buffer. Normalisation is a similarity
    /// transform so it would not change an angle, but reporting on the raw
    /// coordinates keeps the numbers obviously in real degrees.
    static func windowSummary(_ window: [[Float]]) -> [JointSummary] {
        guard window.count >= 3 else { return [] }
        return PoseLandmarks.angleTriplets.map { triplet in
            let series = jointAngleSeries(window: window, triplet: triplet)
            return JointSummary(joint: triplet.name,
                                rom: rangeOfMotion(series),
                                tremor: tremorMetrics(series))
        }
    }
}
